File: app/src/main/java/com/cxwl/shawn/wuzhishan/decision/activity/PDFActivity.java
package com.cxwl.shawn.wuzhishan.decision.activity;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.os.Message;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.cxwl.shawn.wuzhishan.decision.R;
import com.cxwl.shawn.wuzhishan.decision.common.CONST;
import com.cxwl.shawn.wuzhishan.decision.util.OkHttpUtil;
import com.github.barteksc.pdfviewer.PDFView;
import com.github.barteksc.pdfviewer.scroll.DefaultScrollHandle;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.Request;
import okhttp3.Response;

/**
 * PDF文档
 */
public class PDFActivity extends BaseActivity implements OnClickListener {

	private LinearLayout llBack;
	private TextView tvTitle,tvPercent;
	private PDFView pdfView;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_pdf);
		initWidget();
		initPDFView();
	}

	private void initWidget() {
		llBack = findViewById(R.id.llBack);
		llBack.setOnClickListener(this);
		tvTitle = findViewById(R.id.tvTitle);
		tvPercent = findViewById(R.id.tvPercent);

		String title = getIntent().getStringExtra(CONST.ACTIVITY_NAME);
		if (!TextUtils.isEmpty(title)) {
			tvTitle.setText(title);
		}
	}

	// 根据Unicode编码完美的判断中文汉字和符号
    private static boolean isChinese(char c) {
        Character.UnicodeBlock ub = Character.UnicodeBlock.of(c);
        if (ub == Character.UnicodeBlock.CJK_UNIFIED_IDEOGRAPHS || ub == Character.UnicodeBlock.CJK_COMPATIBILITY_IDEOGRAPHS
                || ub == Character.UnicodeBlock.CJK_UNIFIED_IDEOGRAPHS_EXTENSION_A || ub == Character.UnicodeBlock.CJK_UNIFIED_IDEOGRAPHS_EXTENSION_B
                || ub == Character.UnicodeBlock.CJK_SYMBOLS_AND_PUNCTUATION || ub == Character.UnicodeBlock.HALFWIDTH_AND_FULLWIDTH_FORMS
                || ub == Character.UnicodeBlock.GENERAL_PUNCTUATION) {
            return true;
        }
        return false;
    }

    // 完整的判断中文汉字和符号
    private String isChinese(String strName) {
        char[] ch = strName.toCharArray();
        for (char c : ch) {
            if (isChinese(c)) {
            	try {
					strName = strName.replace(c+"", URLEncoder.encode(c+"", "UTF-8"));
				} catch (UnsupportedEncodingException e) {
					e.printStackTrace();
				}
            }
        }
		return strName;
    }

	private void initPDFView() {
		pdfView = findViewById(R.id.pdfView);
		pdfView.enableDoubletap(true);
		pdfView.enableSwipe(true);

		String pdfUrl = getIntent().getStringExtra(CONST.WEB_URL);
		if (TextUtils.isEmpty(pdfUrl)) {
			return;
		}else {
			pdfUrl = isChinese(pdfUrl);
		}

		OkHttpFile(pdfUrl);
	}

	private void OkHttpFile(final String url) {
		new Thread(new Runnable() {
			@Override
			public void run() {
				OkHttpUtil.enqueue(new Request.Builder().url(url).build(), new Callback() {
					@Override
					public void onFailure(Call call, IOException e) {

					}

					@Override
					public void onResponse(Call call, Response response) throws IOException {
						if (!response.isSuccessful()) {
							return;
						}
						InputStream is = null;
						FileOutputStream fos = null;
						try {
							is = response.body().byteStream();//获取输入流
							float total = response.body().contentLength();//获取文件大小
							if(is != null){
								File files = new File(Environment.getExternalStorageDirectory()+"/Wuzhishan");
								if (!files.exists()) {
									files.mkdirs();
								}
								String filePath = files.getAbsolutePath()+"/"+"1.pdf";
								fos = new FileOutputStream(filePath);
								byte[] buf = new byte[1024];
								int ch = -1;
								int process = 0;
								while ((ch = is.read(buf)) != -1) {
									fos.write(buf, 0, ch);
									process += ch;

									int percent = (int) Math.floor((process / total * 100));
									Log.e("percent", process+"--"+total+"--"+percent);
									Message msg = handler.obtainMessage(1001);
									msg.what = 1001;
									msg.obj = filePath;
									msg.arg1 = percent;
									handler.sendMessage(msg);

								}

							}
							fos.flush();
							fos.close();// 下载完成

						} catch (Exception e) {
							e.printStackTrace();
						} finally {
							if (is != null) {
								is.close();
							}
							if (fos != null) {
								fos.close();
							}
						}

					}
				});
			}
		}).start();
	}

	@SuppressLint("HandlerLeak")
	private Handler handler = new Handler() {
		public void handleMessage(android.os.Message msg) {
			if (msg.what == 1001) {
				if (tvPercent == null || pdfView == null) {
					return;
				}
				int percent = msg.arg1;
				tvPercent.setText(percent+getString(R.string.unit_percent));
				if (percent >= 100) {
					tvPercent.setVisibility(View.GONE);
					String filePath = msg.obj+"";
					if (!TextUtils.isEmpty(filePath)) {
						File file = new File(msg.obj+"");
						if (file.exists()) {
							pdfView.fromFile(file)
									.defaultPage(0)
									.scrollHandle(new DefaultScrollHandle(PDFActivity.this))
									.load();
						}
					}
				}
			}
		};
	};

	@Override
	public void onClick(View v) {
		switch (v.getId()) {
		case R.id.llBack:
			finish();
			break;

		default:
			break;
		}
	}
}

File: app/src/main/java/com/cxwl/shawn/wuzhishan/decision/activity/PDFActivity.kt
package com.cxwl.shawn.wuzhishan.decision.activity

import android.Manifest
import android.annotation.SuppressLint
import android.content.pm.PackageManager
import android.os.*
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.text.TextUtils
import android.view.View
import android.view.View.OnClickListener
import com.cxwl.shawn.wuzhishan.decision.R
import com.cxwl.shawn.wuzhishan.decision.common.CONST
import com.cxwl.shawn.wuzhishan.decision.util.AuthorityUtil
import com.cxwl.shawn.wuzhishan.decision.util.OkHttpUtil
import com.github.barteksc.pdfviewer.scroll.DefaultScrollHandle
import kotlinx.android.synthetic.main.activity_pdf.*
import kotlinx.android.synthetic.main.layout_title.*
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Request
import okhttp3.Response
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

/**
 * PDF文档
 */
class PDFActivity : BaseActivity(), OnClickListener {

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_pdf)
		initWidget()
		checkAuthority()
	}

	private fun initWidget() {
		llBack.setOnClickListener(this)

		val title = intent.getStringExtra(CONST.ACTIVITY_NAME)
		if (!TextUtils.isEmpty(title)) {
			tvTitle.text = title
		}
	}

	private fun initPDFView() {
		pdfView.enableDoubletap(true)
		pdfView.enableSwipe(true)

		val pdfUrl = intent.getStringExtra(CONST.WEB_URL)
		OkHttpFile(pdfUrl)
	}

	private fun OkHttpFile(url : String) {
		Thread(Runnable {
			OkHttpUtil.enqueue(Request.Builder().url(url).build(), object : Callback {
				override fun onFailure(call: Call, e: IOException) {
				}
				override fun onResponse(call: Call, response: Response) {
					if (!response.isSuccessful) {
						return
					}
					val totalLength = response.body!!.contentLength()//获取文件大小
					val inputStream = response.body!!.byteStream()//获取输入流
					val buffer = ByteArray(1024)
					val files = File(Environment.getExternalStorageDirectory().toString()+"/HainanNQ")
					if (!files.exists()) {
						files.mkdirs()
					}
					val filePath = files.absolutePath+"/"+"1.pdf"
					var len = 0
					var sum = 0

					val fos = FileOutputStream(filePath)
					while (inputStream.read(buffer).apply { len = this } > 0) {
						fos.write(buffer, 0, len)
						sum += len

						val percent = (sum * 100 / totalLength).toInt()
						val msg = handler.obtainMessage(1001)
						msg.what = 1001
						msg.obj = filePath
						msg.arg1 = percent
						handler.sendMessage(msg)
					}
					fos.flush()
					fos.close()// 下载完成

				}
			})
		}).start()
	}

	private val handler : Handler = @SuppressLint("HandlerLeak")
	object : Handler() {
		override fun handleMessage(msg: Message?) {
			super.handleMessage(msg)
			when(msg!!.what) {
				1001 -> {
					if (tvPercent == null || pdfView == null) {
						return
					}
					val percent = msg.arg1
					tvPercent.text = percent.toString()+getString(R.string.unit_percent)
					if (percent >= 100) {
						tvPercent.visibility = View.GONE
						val filePath = msg.obj.toString()
						if (!TextUtils.isEmpty(filePath)) {
							val file = File(filePath)
							if (file.exists()) {
								pdfView.fromFile(file)
										.defaultPage(0)
										.scrollHandle(DefaultScrollHandle(this@PDFActivity))
										.load()
							}
						}
					}
				}
			}
		}
	}

	override fun onClick(view: View?) {
		when(view!!.id) {
			R.id.llBack -> finish()
		}
	}

	/**
	 * 申请权限
	 */
	private fun checkAuthority() {
		if (Build.VERSION.SDK_INT < 23) {
			initPDFView()
		} else {
			val checkSelfPermission = ContextCompat.checkSelfPermission(this@PDFActivity,
					Manifest.permission.WRITE_EXTERNAL_STORAGE)
			if (checkSelfPermission == PackageManager.PERMISSION_GRANTED) {
				initPDFView()
			} else {
				ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), AuthorityUtil.AUTHOR_STORAGE)
			}
		}
	}

	override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
		super.onRequestPermissionsResult(requestCode, permissions, grantResults)
		when (requestCode) {
			AuthorityUtil.AUTHOR_STORAGE -> if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
				initPDFView()
			} else {
				if (!ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
					AuthorityUtil.intentAuthorSetting(this, "\"" + getString(R.string.app_name) + "\"" + "需要使用存储权限，是否前往设置？")
				}
			}
		}
	}

}
